### PR TITLE
Add ethereum transact call dispatch

### PR DESCRIPTION
**Pull request by AsceticBear**
This is an early draft pull request in development. Because of the cycle dep problem, the concrete tests have to be done on `darwinia-common` side.

### CODE REVIEW

**Comment by wuminzhe** [comment]
I suggest renaming EthereumCallDispatch trait to CallDispatch.  the CallDispatch is used to dispatch different calls.

File: modules/dispatch/src/lib.rs [hunk 1]
```diff
@@ -297,6 +311,39 @@ impl<T: Config<I>, I: 'static> MessageDispatch<T::AccountId, T::BridgeMessageId>
 			return dispatch_result
 		}
 
+		// Dispatch Ethereum call earlier before pay
+		match T::EthereumCallDispatcher::dispatch(&call, &origin_account) {
```
Comment (wuminzhe) [inline modules/dispatch/src/lib.rs:315]: I feel like this is a patch here. 
maybe, generalize the `EthereumCallDispatch` to `CallDispatch`.

or, something like:
replace 
`let origin = RawOrigin::Signed(origin_account).into();` 
with
`let origin = T::OriginBuilder.getOrigin(call, origin_account);`

Comment (AsceticBear) [reply modules/dispatch/src/lib.rs:317]: > generalize the EthereumCallDispatch to CallDispatch.

It seems like a rename issue. For dispatch calls besides the Ethereum pallet, we use the original call dispatch in this pallet. The dispatcher I added is only used for the Ethereum pallet dispatch call, so I think `EthereumCallDispatch` is more suitable here.

**Comment by AsceticBear** [comment]
@HackFisher @AurevoirXavier PTAR

File: modules/dispatch/src/lib.rs (hunk 1 above)
Comment (AsceticBear) [reply modules/dispatch/src/lib.rs:317]: Currently, this one is only used by the Ethereum pallet. I don't see any other pallet that needs it. Rename to `CallDispatch` is wired this pallet already called `pallet-messages-dispatch`, then there is another associated type called `CallDispatch`, After the CallDispatch executes logic, there is another call dispatch.

File: modules/dispatch/src/lib.rs [hunk 2]
```diff
@@ -134,6 +146,10 @@ pub mod pallet {
 		),
 		/// Message has been dispatched with given result.
 		MessageDispatched(ChainId, BridgeMessageIdOf<T, I>, DispatchResult),
+		EthereumCallDispatched(ChainId, BridgeMessageIdOf<T, I>, DispatchResult),
+		EthereumCallValidityError(ChainId, BridgeMessageIdOf<T, I>, EthereumCallValidityError),
```
Comment (HackFisher) [inline modules/dispatch/src/lib.rs:150]: This is too specific to put them in this generic pallet.

File: modules/dispatch/src/lib.rs (hunk 1 above)
Comment (HackFisher) [reply modules/dispatch/src/lib.rs:317]: I agree with @wuminzhe , why not use `CallFilter` for validation, and do the duplicate dispatch and cases match  which will be called later.

Why not directly use the spec described in https://github.com/darwinia-network/darwinia-messages-substrate/issues/100 ?

**Review by HackFisher — CHANGES_REQUESTED** [review]

**Comment by AsceticBear** [comment]
Prefer https://github.com/darwinia-network/darwinia-messages-substrate/pull/111

**Pull request closed without merging**